### PR TITLE
Quizmaster

### PR DESCRIPTION
**Pull request by FuroYT**

### CODE REVIEW

**Comment by Ultradragon005** [comment]
holy shid

**Comment by FuroYT** [comment]
finally github isnt shitting itself anymore

**Comment by FuroYT** [comment]
tier 2 progress is for tomorrow ;)

**Comment by spuhnj** [comment]
yippeeee

**Comment by 0xDrMoe** [comment]
@FuroYT ETA? If not ready, this will be in next dev

**Comment by FuroYT** [comment]
I will try to finish it today and tmr

**Comment by FuroYT** [comment]
But before that I'll need to fix conflicts (thanks GitHub)

File: Modules/OptionHolder.cs [hunk 1]
```diff
@@ -2970,17 +2968,16 @@ public static void Load()
...
         //HighLevelAntiCheat = StringOptionItem.Create(60260, "HighLevelAntiCheat", CheatResponsesName, 0, TabGroup.SystemSettings, false)
         //.SetHeader(true);
-
         AutoDisplayKillLog = BooleanOptionItem.Create(60270, "AutoDisplayKillLog", true, TabGroup.SystemSettings, false)
             .SetHeader(true);
         AutoDisplayLastRoles = BooleanOptionItem.Create(60280, "AutoDisplayLastRoles", true, TabGroup.SystemSettings, false);
         AutoDisplayLastResult = BooleanOptionItem.Create(60290, "AutoDisplayLastResult", true, TabGroup.SystemSettings, false);
-
+        HideExileChat = BooleanOptionItem.Create(60295, "HideExileChat", true, TabGroup.SystemSettings, false)
+            .SetHeader(true);
```
Comment (Tommy-XL) [inline Modules/OptionHolder.cs:2980]: Extra option

File: Modules/OptionHolder.cs [hunk 2]
```diff
@@ -3135,27 +3116,11 @@ public static void Load()
...
-        DecontaminationTimeOnMiraHQ = FloatOptionItem.Create(60504, "DecontaminationTimeOnMiraHQ", new(0.5f, 10f, 0.25f), 3f, TabGroup.GameSettings, false)
-            .SetParent(ChangeDecontaminationTime)
-            .SetValueFormat(OptionFormat.Seconds)
-            .SetColor(new Color32(19, 188, 233, byte.MaxValue));
-        // Decontamination time on Polus
-        DecontaminationTimeOnPolus = FloatOptionItem.Create(60505, "DecontaminationTimeOnPolus", new(0.5f, 10f, 0.25f), 3f, TabGroup.GameSettings, false)
-            .SetParent(ChangeDecontaminationTime)
-            .SetValueFormat(OptionFormat.Seconds)
-            .SetColor(new Color32(19, 188, 233, byte.MaxValue));
-
```
Comment (Tommy-XL) [inline Modules/OptionHolder.cs:3152]: ... Why was it deleted?

File: Modules/Utils.cs [hunk 3]
```diff
@@ -1954,8 +1954,8 @@ public static void ApplySuffix(PlayerControl player)
                     _ => name
                 };
             }
-            
-            if (!name.Contains('\r') && player.FriendCode.GetDevUser().HasTag() && (player.AmOwner || player.IsModClient()))
+
+            if (!name.Contains($"\r\r") && player.FriendCode.GetDevUser().HasTag())
```
Comment (Tommy-XL) [inline Modules/Utils.cs:1958]: Same here

File: Patches/ChatCommandPatch.cs [hunk 4]
```diff
@@ -2215,6 +2228,41 @@ public static void OnReceiveChat(PlayerControl player, string text, out bool can
                 break;
         }
     }
+    private static void QuizmasterAnswer(PlayerControl plr, string[] args)
+    {
+        if (Quizmaster.MarkedPlayer == plr.PlayerId)
```
Comment (Tommy-XL) [inline Patches/ChatCommandPatch.cs:2233]: Why not move this to `Quizmaster.cs`

File: Modules/OptionHolder.cs [hunk 5]
```diff
@@ -3509,8 +3462,15 @@ public static void Load()
...
+        DisablePolishGem = BooleanOptionItem.Create(60637, "DisablePolishGem", false, TabGroup.TaskSettings, false)
+            .SetParent(DisableCommonTasks)
+            .SetGameMode(CustomGameMode.Standard);
+        DisableHelpCritter = BooleanOptionItem.Create(60638, "DisableHelpCritter", false, TabGroup.TaskSettings, false)
+            .SetParent(DisableCommonTasks)
+            .SetGameMode(CustomGameMode.Standard);
+        DisableHoistSupplies = BooleanOptionItem.Create(60639, "DisableHoistSupplies", false, TabGroup.TaskSettings, false)
+            .SetParent(DisableCommonTasks)
+            .SetGameMode(CustomGameMode.Standard);
```
Comment (Tommy-XL) [inline Modules/OptionHolder.cs:3473]: Extra options

File: Patches/MeetingHudPatch.cs [hunk 6]
```diff
@@ -1515,6 +1515,7 @@ public static void Postfix()
     {
         MeetingStates.FirstMeeting = false;
         Logger.Info("------------End Meeting------------", "Phase");
+        Quizmaster.OnMeetingEnd();
         if (AmongUsClient.Instance.AmHost)
         {
```
Comment (Tommy-XL) [inline Patches/MeetingHudPatch.cs:1520]: Isn't it better to move `Quizmaster.OnMeetingEnd();` after `if (AmongUsClient.Instance.AmHost)`

File: Patches/PlayerControlPatch.cs [hunk 7]
```diff
@@ -1337,6 +1341,8 @@ public static bool Prefix(PlayerControl __instance, [HarmonyArgument(0)] PlayerC
     }
     public static void Postfix(PlayerControl __instance, [HarmonyArgument(0)] PlayerControl target, [HarmonyArgument(1)] MurderResultFlags resultFlags)
     {
+        Quizmaster.OnPlayerDead(target);
+
         if (target.AmOwner) RemoveDisableDevicesPatch.UpdateDisableDevices();
         if (!target.Data.IsDead || !AmongUsClient.Instance.AmHost) return;
```
Comment (Tommy-XL) [inline Patches/PlayerControlPatch.cs:1347]: Same here

File: Modules/ExtendedPlayerControl.cs [hunk 8]
```diff
@@ -741,7 +744,7 @@ CustomRoles.ChiefOfPolice or
             CustomRoles.Killer => true,
 
             _ => pc.Is(CustomRoleTypes.Impostor),
-        };
+        }; ;
```
Comment (Tommy-XL) [inline Modules/ExtendedPlayerControl.cs:747]: Build Error?

**Review by Tommy-XL — CHANGES_REQUESTED** [review]
So

**Comment by FuroYT** [comment]
> So

So um GitHub did shit I think I just pulled the upstream

File: Modules/Utils.cs (hunk 3 above)
Comment (FuroYT) [reply Modules/Utils.cs:1958]: this was intentional so it shows the tag even if you haven't the mod running

**Comment by FuroYT** [comment]
So i've finished i will ask for code review to check if i ain't broke nuthin

**Comment by FuroYT** [comment]
> @FuroYT ETA? If not ready, this will be in next dev

It is ready, now just waiting for review from Tommy for him to check

**Comment by Tommy-XL** [comment]
> It is ready, now just waiting for review from Tommy for him to check

I'll check the code later

**Comment by FuroYT** [comment]
> > It is ready, now just waiting for review from Tommy for him to check
> 
> I'll check the code later

np :)

File: Modules/CustomRolesHelper.cs [hunk 9]
```diff
@@ -260,6 +260,7 @@ public static RoleTypes GetDYRole(this CustomRoles role) // Role has a kill butt
             CustomRoles.Agitater => RoleTypes.Impostor,
             CustomRoles.Spiritcaller => RoleTypes.Impostor,
             CustomRoles.ChiefOfPolice => RoleTypes.Impostor,
+            CustomRoles.Quizmaster => RoleTypes.Impostor,
```
Comment (Gurge44) [inline Modules/CustomRolesHelper.cs:263]: This function is used to make desync impostor roles (neutrals and crewmates with impostor basis) and you don't need to add normal impostor roles to this

**Comment by Tommy-XL** [comment]
Need fix json again :hehe_sob:

**Comment by Tommy-XL** [comment]
And roleColor.json

**Comment by FuroYT** [comment]
> Need fix json again :hehe_sob: 

Can you just get the latest changes from the json and put it cuz I don't know how to fix this one out

**Comment by Tommy-XL** [comment]
> Can you just get the latest changes from the json and put it cuz I don't know how to fix this one out

You can undo all changes to .json and add from via github
image

**Comment by FuroYT** [comment]
> You can undo all changes to .json and add from via github
> 
> image
> 
> 

I'll do it when I get home

**Comment by 0xDrMoe** [comment]
Please resolve this. Let's get it into this dev version.

**Comment by FuroYT** [comment]
> Please resolve this. Let's get it into this dev version.

I’ll do it as quick as I can with school and personal life :]

**Comment by FuroYT** [comment]
@Tommy-XL Finished the json void problem :]

**Comment by 0xDrMoe** [comment]
Is this ready then?

File: Modules/OptionHolder.cs [hunk 10]
```diff
@@ -3293,12 +3271,10 @@ public static void Load()
...
             .SetParent(LightsOutSpecialSettings)
             .SetGameMode(CustomGameMode.Standard);
-
-
-        // Disable
+        //禁用相关设定
         TextOptionItem.Create(10000027, "MenuTitle.Disable", TabGroup.GameSettings)
-            .SetColor(new Color32(255, 153, 153, byte.MaxValue))
-            .HideInHnS();
+            .SetGameMode(CustomGameMode.Standard)
```
Comment (Tommy-XL) [inline Modules/OptionHolder.cs:3276]: This should be hidden for H&S

File: Modules/OptionHolder.cs [hunk 11]
```diff
@@ -3316,19 +3292,17 @@ public static void Load()
...
             .SetGameMode(CustomGameMode.Standard)
             .SetColor(new Color32(255, 153, 153, byte.MaxValue));
         DisableCloseDoor = BooleanOptionItem.Create(60566, "DisableCloseDoor", false, TabGroup.GameSettings, false)
             .SetGameMode(CustomGameMode.Standard)
             .SetColor(new Color32(255, 153, 153, byte.MaxValue));
-
         // Disable Devices
         DisableDevices = BooleanOptionItem.Create(60570, "DisableDevices", false, TabGroup.GameSettings, false)
-            .SetColor(new Color32(255, 153, 153, byte.MaxValue))
-            .HideInHnS();
```
Comment (Tommy-XL) [inline Modules/OptionHolder.cs:3331]: Same

File: Modules/OptionHolder.cs [hunk 12]
```diff
@@ -3568,15 +3545,15 @@ public static void Load()
         DisableDivertPower = BooleanOptionItem.Create(60669, "DisableDivertPower", false, TabGroup.TaskSettings, false)
             .SetParent(DisableOtherTasks);
         DisableActivateWeatherNodes = BooleanOptionItem.Create(60670, "DisableActivateWeatherNodes", false, TabGroup.TaskSettings, false)
-            .SetParent(DisableOtherTasks);
+            .SetParent(DisableOtherTasks)
+            .SetGameMode(CustomGameMode.Standard);
```
Comment (Tommy-XL) [inline Modules/OptionHolder.cs:3549]: Remove .`SetGameMode(CustomGameMode.Standard);`

File: Modules/OptionHolder.cs [hunk 13]
```diff
@@ -3551,7 +3527,8 @@ public static void Load()
         DisableHelpCritter = BooleanOptionItem.Create(60638, "DisableHelpCritter", false, TabGroup.TaskSettings, false)
             .SetParent(DisableLongTasks);
         DisableHoistSupplies = BooleanOptionItem.Create(60639, "DisableHoistSupplies", false, TabGroup.TaskSettings, false)
-            .SetParent(DisableLongTasks);
+            .SetParent(DisableLongTasks)
+            .SetGameMode(CustomGameMode.Standard);
```
Comment (Tommy-XL) [inline Modules/OptionHolder.cs:3531]: Same

**Review by Tommy-XL — APPROVED** [review]

**Pull request merged**